### ISSUE TITLE
NaN gradient when using batch normalization

### ISSUE THREAD

**Issue by MaigoAkisame** [Theano/Theano#5546]
I was building a neural network with batch normalization layers using Keras and Theano 0.9.0b1. I found that the batch normalization layer always produced NaN gradients when the dimension to keep was not dimension 1 (0-based). A minimal example that produces the problem can be found here:

https://github.com/fchollet/keras/issues/5377

I had to switch back to Theano 0.8.2 to avoid the problem :(

**Comment by lamblin** [comment]
I just tried it with the latest master, and I could reproduce the issue to some extent (no NaN, but values that were different for CPU and GPU), and #5552 seems to fix it.

**Comment by lamblin** [comment]
#5552 has been merged now, could you try with the latest master?
Thanks!

**Comment by MaigoAkisame** [comment]
Thanks! Currently I have some jobs running with the old version of Theano; I'll try with the latest version when they finish (maybe in 2 days).

**Comment by MaigoAkisame** [comment]
Oops, I still get NaN with the latest version...

**Comment by lamblin** [comment]
Still with the same script?
Could you try to identify a seed for `numpy.random` that produces the `NaN` reliably?
Thanks

**Comment by MaigoAkisame** [comment]
In fact any seed will do...

For example, I set `numpy.random.seed(15213)` before the line `x = ...`, and I got the following output:
```
[(bn_gamma, array([ nan], dtype=float32)), (bn_beta, array([ 0.02858536], dtype=float32))]
```

By the way, here's my `~/.theanorc` file:
```
[global]
floatX = float32
device = gpu
[nvcc]
fastmath = True
[lib]
cnmem = 1
[cuda]
root = /usr/local/cuda-8.0
```

And here's my ~/.keras/keras.json file:
```
{
    "image_dim_ordering": "tf",
    "epsilon": 1e-07,
    "floatx": "float32",
    "backend": "theano"
}
```

**Comment by lamblin** [comment]
Thanks, I'll try to reproduce.

**Comment by lamblin** [comment]
OK, so with `nvcc.fastmath,device=gpu`, I can reproduce the issue:
```
[(bn_gamma, array([ nan], dtype=float32)), (bn_beta, array([ 0.02858536], dtype=float32))]
```

However, keeping the default for `nvcc.fastmath`, or using the new back-end which ignores the `nvcc` section works:
```
[(bn_gamma, array([ 0.18067236], dtype=float32)), (bn_beta, array([ 0.02858542], dtype=float32))]
```

It is not really surprising, since fastmath can collapse small values to 0.

How much of a speed-up do you typically gain with fastmath?

**Comment by MaigoAkisame** [comment]
I haven't compared the speed with and without fastmath. I have had that option there ever since I started to learn Theano, because at least the name indicates things may get faster.

Would you recommend that I disable fastmath?

**Comment by lamblin** [comment]
If it is the cause of numerical instability, then yes.
Actually, I recommend you switch to the new back-end for a large number of reasons, and the new back-end does not support fastmath :)

Since I don't think there's anything to fix in Theano itself, I'll close this issue.
